zpages/spanprocessor_test.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2017, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zpages
+
+import (
+	"context"
+	"reflect"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/codes"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+func TestSpanProcessor_DoNothing(t *testing.T) {
+	zsp := NewSpanProcessor()
+	assert.NoError(t, zsp.ForceFlush(context.Background()))
+	assert.NoError(t, zsp.Shutdown(context.Background()))
+}
+
+func TestSpanProcessor(t *testing.T) {
+	zsp := NewSpanProcessor()
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithSpanProcessor(zsp),
+	)
+
+	const spanName = "testSpan"
+	const numSpans = 9
+
+	tracer := tracerProvider.Tracer("test")
+	spans := createActiveSpans(tracer, spanName, numSpans)
+	// Sort the spans by the address pointer so we can compare.
+	sort.Slice(spans, func(i, j int) bool {
+		return reflect.ValueOf(spans[i]).Pointer() < reflect.ValueOf(spans[j]).Pointer()
+	})
+	require.Len(t, spans, numSpans)
+	activeSpans := zsp.activeSpans(spanName)
+	assert.Len(t, activeSpans, numSpans)
+	// Sort the activeSpans by the address pointer so we can compare.
+	sort.Slice(activeSpans, func(i, j int) bool {
+		return reflect.ValueOf(activeSpans[i]).Pointer() < reflect.ValueOf(activeSpans[j]).Pointer()
+	})
+	for i := range spans {
+		assert.Same(t, spans[i], activeSpans[i])
+	}
+	// No ended spans so there will be no error, no latency samples.
+	assert.Len(t, zsp.errorSpans(spanName), 0)
+	for i := 0; i < defaultBoundaries.numBuckets(); i++ {
+		assert.Len(t, zsp.spansByLatency(spanName, i), 0)
+	}
+	spansPM := zsp.spansPerMethod()
+	require.Equal(t, 1, len(spansPM))
+	assert.Equal(t, numSpans, spansPM[spanName].activeSpans)
+
+	// End all Spans, they will end pretty fast, so we can only check that we have at least one in
+	// errors and one in latency samples.
+	for _, s := range spans {
+		s.End()
+	}
+	// Test that no more active spans.
+	assert.Len(t, zsp.activeSpans(spanName), 0)
+	assert.LessOrEqual(t, 1, len(zsp.errorSpans(spanName)))
+	numLatencySamples := 0
+	for i := 0; i < defaultBoundaries.numBuckets(); i++ {
+		numLatencySamples += len(zsp.spansByLatency(spanName, i))
+	}
+	assert.LessOrEqual(t, 1, numLatencySamples)
+}
+
+func TestSpanProcessor_Fuzzer(t *testing.T) {
```

Comment:
Can you explain a bit what this test is doing, and how it can fail?

It looks like it might fail if there is only the 2>= spansPerMethod().   Isn't `len()` always going to be 0 or more?

Reply:
As the name said I am not testing necessarily functionality here but to ensure that multi threading works and no race conditions are detected